portalocker/portalocker.py

```diff
@@ -104,14 +107,28 @@ def lock(file_: typing.Union[typing.IO, int], flags: LockFlags):
             )
 
         try:
-            fcntl.flock(file_, flags)
-        except locking_exceptions as exc_value:
-            # The exception code varies on different systems so we'll catch
-            # every IO error
-            raise exceptions.LockException(exc_value, fh=file_) from exc_value
+            LOCKER(file_, flags)
+        except OSError as exc_value:
+            # Python can use one of several different exception classes to represent
+            # timeout (most likely is BlockingIOError and IOError), but these errors
+            # may also represent other failures. On some systems, `IOError is OSError`
+            # which means checking for either IOError or OSError can mask other errors.
+            # The safest check is to catch OSError (from which the others inherit)
+            # and check the errno (which should be EACCESS or EAGAIN according to the
+            # spec).
+            if exc_value.errno in (errno.EACCES, errno.EAGAIN):
+                # A timeout exception, wrap this so the outer code knows to try again
+                # (if it wants to).
+                # TODO: Would AlreadyLocked by a better error to raise here? Or perhaps
+                # a new exception class like TryAgain?
+                raise exceptions.LockException(exc_value, fh=file_) from exc_value
+            else:
+                # Something else went wrong; don't wrap this so we stop immediately.
+                raise
```

Comment:
I'm not too keen on implementing this since it could break backwards compatibility. I'm not sure what kind of errors can pop up at all platforms so it should always be wrapped in a `LockException` to be compatible with the rest of the code and not break existing implementations

Reply:
Fair enough, generally I prefer not to wrap errors and just re-raise but here I think it could make sense, especially if it's already established functionality.

The main thing is to raise a different exception for timeout as opposed to any other exception, so the outer loop only tries again for timeouts. Otherwise a non-timeout exception (like giving the wrong syntax to `fcntl`) won't propagate until the end of the timeout. For an infinite timeout (which would be unwise anyway but is possible) the loop would never exit otherwise...